src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -245,11 +250,20 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private Activity StartActivity(HttpContext httpContext, out bool hasDiagnosticListener)
+        private Activity? StartActivity(HttpContext httpContext, bool loggingEnabled, bool diagnosticListenerActivityCreationEnabled, out bool hasDiagnosticListener)
         {
-            var activity = new Activity(ActivityName);
+            var activity = _activitySource.CreateActivity(ActivityName, ActivityKind.Server);
+            if (activity is null && (loggingEnabled || diagnosticListenerActivityCreationEnabled))
+            {
+                activity = new Activity(ActivityName);
+            }
             hasDiagnosticListener = false;
 
+            if (activity is null)
+            {
+                return null;
```

Comment:
Review: I don't think this code is reachable and I can drop this null check

Reply:
this can be reached in very corner case. can be before calling StartActivity, you had loggingEnabled = false and diagnosticListenerActivityCreationEnabled = false, and _activitySource.HasListeners() was true. then when started to execute StartActivity, somehow the ActivitySource listener get unregistered. I know this is unlikely scenario but it is possible :-)

---
In reply to: [581423911](https://github.com/dotnet/aspnetcore/pull/30089#discussion_r581423911) [](ancestors = 581423911)

Reply:
Ahh, that's right! I'm going to keep this check 😄